### PR TITLE
[UNR-4548] Run tests using the replication graph nightly

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Change CI so that it runs tests with the replication graph at a nightly cadence.

Also fixed or otherwise dealt with the tests that would have failed as a result of this.

Copy-pasted dev notes:
```
Fail with native replication graph:
 - SpatialTestTearOff
   - TearOff can't be called on BeginPlay in RepGraph
   - Fix: move to first tick
 - SpatialTestInitialOnlyForInterestActor
   - Test had a race condition
   - Fix: fix the race condition
 - SpatialTestInitialOnlyForSpawnActor
   - Test had a race condition
   - Fix: fix the race condition
 - VisibilityTest
   - Test just doesn't make much sense with replication graph
   - Fix: special-case it so that it tests our expectation for replication graph
 - SpatialTestNetOwnership
   - Warning is not produced with native replication graph
   - It seemingly IS produced with spatial replication graph - ???
   - Trying to fix this breaks the test framework - ?!?
   - NotReallyAFix: The test just doesn't run with replication graph.
     Added a bunch of ticket references to fix this in the future.
     Really bad.

Fail with spatial replication graph:
 - SpatialTestPlayerControllerHandover
   - Only fails with spatial replication graph
   - Cause: the debug interface didn't work with the replication graph
   - Fix: Fix the debug interface so now it ticks correctly with the replication graph (maybe)
 - SpatialComponentTest
   - Can now crash with spatial replication graph
   - Fix: fix race condition in the test
```

#### Release note
Not required, I think.

#### Tests
Slow test run (green): https://buildkite.com/improbable/unrealgdk-premerge/builds/23514#_

#### Documentation
Hmmmmm, interesting question. I want people to know that adding these setting toggles to our CI suites is quite easy, not sure where to best communicate that...

#### Primary reviewers
@m-samiec @ImprobableNic + someone else? Dmitri?

### CODE REVIEW

**Comment by mironec** [comment]
Limitation: this does not work in Mac CI. I could have made it work, but it would be not as simple as the Windows change, and would require some code duplication.

Since we will be moving to the new CI system at some point, and since Mac shouldn't really affect core networking functionality, I decided to leave it out.

**Comment by mironec** [comment]
Also a manual run with initial only enabled: https://buildkite.com/improbable/unrealgdk-premerge/builds/23517

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -1851,6 +1851,11 @@ int32 USpatialNetDriver::ServerReplicateActors(float DeltaSeconds)
 	}
 	check(SpatialConnection->bReliableSpatialConnection);
 
+	if (DebugCtx != nullptr)
+	{
+		DebugCtx->TickServer();
+	}
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1857]: @ImprobableNic or @improbable-dmitrii what's your opinion on moving it here?
The repGraph returns on the next line so the debug interface was not running. I think it's safe to move it up here, but you will have more context.

File: ci/setup-build-test-gdk.ps1 [hunk 2]
```diff
@@ -79,6 +79,10 @@ if ((Test-Path env:TEST_CONFIG) -And ($env:TEST_CONFIG -eq "Native")) {
         $tests[0].tests_path += "+${test_map_path}CI_Nightly/"
         $tests[0].test_results_dir = "Slow" + $tests[0].test_results_dir
 
+        # We run functional spatial tests against Vanilla UE4 with replication graph enabled
+        $tests += [TestSuite]::new($gdk_test_project, "SpatialNetworkingMap", "VanillaTestResultsRepGraph", "${test_map_path}CI_Premerge/+${test_map_path}CI_Nightly/",
+        "$user_gdk_settings", $False, "-ini:Engine:[/Script/OnlineSubsystemUtils.IpNetDriver]:ReplicationDriverClassName=/Script/GDKTestGyms.TestGymsReplicationGraph ${user_cmd_line_args}")
+
```
Comment (mironec) [inline ci/setup-build-test-gdk.ps1:85]: This and the lines below are the only things needed to add a new test suite using a custom setting.
This will change with the new CI system, but should be similarly "easy".

Comment (mironec) [reply ci/setup-build-test-gdk.ps1:85]: at least it will have named arguments in Python

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp [hunk 3]
```diff
@@ -115,10 +115,11 @@ void ASpatialTestInitialOnlyForInterestActor::PrepareTest()
 	AddStep(TEXT("Server change value."), FWorkerDefinition::Server(1), nullptr, [this]() {
 		TArray<AActor*> SpawnActors;
 		UGameplayStatics::GetAllActorsOfClass(GetWorld(), ASpatialTestInitialOnlySpawnActor::StaticClass(), SpawnActors);
+		AssertEqual_Int(SpawnActors.Num(), 1, TEXT("There should be exactly one InitialOnly actor in the world."));
 		for (AActor* Actor : SpawnActors)
 		{
 			ASpatialTestInitialOnlySpawnActor* SpawnActor = Cast<ASpatialTestInitialOnlySpawnActor>(Actor);
-			if (SpawnActor != nullptr)
+			if (AssertIsValid(SpawnActor, TEXT("SpawnActor should be valid.")))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp:122]: Did this actual fail? Or just a driveby?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp [hunk 4]
```diff
@@ -128,16 +129,17 @@ void ASpatialTestInitialOnlyForInterestActor::PrepareTest()
...
 			{
-				AssertTrue(SpawnActor->Int_Initial == 1, TEXT("Check Actor.Int_Initial value."));
-				AssertTrue(SpawnActor->Int_Replicate == 2, TEXT("Check Actor.Int_Replicate value."));
+				RequireEqual_Int(SpawnActor->Int_Initial, 1, TEXT("Check Actor.Int_Initial value."));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp:141]: I'm not sure this should be a require - it should always be 1. If the require passes on the first tick (because it just hasn't replicated), will it actually fail on a subsequent tick if the initial only value does get updated?
Perhaps switch to an assert check that fires on when the INT_replicate value has been updated to 2.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp [hunk 5]
```diff
@@ -103,36 +104,24 @@ void ASpatialTestInitialOnlyForSpawnActor::PrepareTest()
...
+			AssertEqual_Int(SpawnActors.Num(), 1, TEXT("There should be exactly one InitialOnly actor in the world."));
 			for (AActor* Actor : SpawnActors)
 			{
 				ASpatialTestInitialOnlySpawnActor* SpawnActor = Cast<ASpatialTestInitialOnlySpawnActor>(Actor);
-				if (SpawnActor != nullptr)
+				if (AssertIsValid(SpawnActor, TEXT("SpawnActor should be valid.")))
 				{
-					AssertTrue(SpawnActor->Int_Initial == 1, TEXT("Check Actor.Int_Initial value."));
-					AssertTrue(SpawnActor->Int_Replicate == 2, TEXT("Check Actor.Int_Replicate value."));
+					RequireEqual_Int(SpawnActor->Int_Initial, 1, TEXT("Check Actor.Int_Initial value."));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:117]: Likewise here

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp [hunk 6]
```diff
@@ -103,36 +104,24 @@ void ASpatialTestInitialOnlyForSpawnActor::PrepareTest()
...
+					RequireEqual_Int(SpawnActor->Int_Replicate, 2, TEXT("Check Actor.Int_Replicate value."));
 				}
 			}
 
 			FinishStep();
-		});
+		},
+		10.0f);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:124]: Agree there should be a timelimit when using `Require`, can you add it to the other tests you modified?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp (hunk 3 above)
Comment (jeffxu-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp:122]: I think this is an actual fail here. `SpawnActor` should have been created in the previous steps.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/ReplicatedTearOffActor.cpp [hunk 7]
```diff
@@ -8,14 +8,26 @@ AReplicatedTearOffActor::AReplicatedTearOffActor()
...
-	TearOff();
+	bFirstTick = true;
+}
+
+void AReplicatedTearOffActor::Tick(float DeltaSeconds)
+{
+	if (bFirstTick)
+	{
+		// Note:  calling TearOff inside the constructor will not prevent the Actor from replicating.
+		// Note2: calling TearOff inside BeginPlay will produce an error if using the ReplicationGraph.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/ReplicatedTearOffActor.cpp:27]: Curious what the error is? And does it produce an error in native replication graph also?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp [hunk 8]
```diff
@@ -67,8 +83,10 @@ void ASpatialTestNetOwnership::PrepareTest()
 
 	// Server 1 spawns the NetOwnershipCube and registers it for auto-destroy.
 	AddStep(TEXT("SpatialTestNetOwnershipServerSpawnCube"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		// The position is chosen as a hack to make sure the cube spawns on Server 1's turf, so we don't run into issues with the framework
+		// itself...
 		ANetOwnershipCube* Cube =
-			GetWorld()->SpawnActor<ANetOwnershipCube>(FVector::ZeroVector, FRotator::ZeroRotator, FActorSpawnParameters());
+			GetWorld()->SpawnActor<ANetOwnershipCube>(FVector(-50.0f, -50.0f, 0.0f), FRotator::ZeroRotator, FActorSpawnParameters());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp:89]: Could use `GetWorkerEntityPosition` here

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp [hunk 9]
```diff
@@ -79,6 +81,11 @@ void AVisibilityTest::PrepareTest()
 
 				FinishStep();
 			}
+
+			if (AssertTrue(HasAuthority(), TEXT("We should have authority over the test actor.")))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp:85]: nit - looks strange having logic run after FinishStep has been called, can we move this to before the IsValid(PC) check?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp [hunk 10]
```diff
@@ -79,6 +81,11 @@ void AVisibilityTest::PrepareTest()
...
+			{
+				bRunningWithReplicationGraph = GetNetDriver()->GetReplicationDriver() != nullptr;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp:87]: Also why replicate this?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp [hunk 11]
```diff
@@ -103,36 +104,24 @@ void ASpatialTestInitialOnlyForSpawnActor::PrepareTest()
...
-				if (SpawnActor != nullptr)
-				{
-					IsReady = true;
-				}
-			}
-			return IsReady;
-		},
-		[this]() {
+		TEXT("Check changed value."), FWorkerDefinition::Client(1), nullptr, nullptr,
+		[this](float DeltaTime) {
```
Comment (jeffxu-improbable) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:108]: Why we use a `TickEvent` here instead?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp:122]: I think Mike was asking whether I have seen it fail in my testing, and I haven't.
But I think my philosophy for writing tests is that for every `if`, I ask what I want the test to do if that condition is unsatisified: in this case, if the SpawnActor is not valid, the test would previously proceed to finish step and probably fail at some point further down. However, not having the SpawnActor is pretty bad thing and I would instead expect the test to fail, so I try replacing flow-control (where possible) with assert-and-flow-control.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp:141]: For the first point, it needs to be a require, because the previous step is setting the values on the server, but this step is checking them on the client. There's no guarantee that the property replication will arrive on the client faster than the RPC which signals the client to start the next step in the test. Usually, we deal with this using the requires to wait until the condition is satisfied.
For your second point, both requires have to be satisfied at the same time for the test to successfully finish the step. So it's kind of an implementation detail, but we will basically wait until Int_Replicate becomes 2, and Int_Initial has to be 1 at the same time. The alternative would be the "wait-so-bad-thing-does-not-happen" pattern, where we could introduce a wait here of e.g. 1 second to "make sure" that the replication for the Int_Initial never comes.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp (hunk 11 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:108]: See my reply to Mike's comment here: https://github.com/spatialos/UnrealGDK/pull/2971/files#r595868053
This is the same kind of thing. If we set a replicated variable in the previous step on server 1, our usual flow for checking whether it has replicated to client 1 in this step would be to have a ticking step which checks, every tick, whether the information has arrived. That is what the `Require`s do - they test the condition supplied to see if it's true, but don't immediately fail if it's not true, the step just can't pass until all requires are satisfied. So in this case, the step will not finish until both Int_Initial == 1 and Int_Replicate == 2. Of course, in case this is not working, the step will never finish. This is why we usually use the timeout together with the requires. When the step times out, the requires will automatically provide debugging info for why the test failed, e.g.: "Check Actor.Int_Initial value. Failed, Received: 2, Expected: 1."

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:108]: And in this case, it also allows us to collapse the ready and start functions into one.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp (hunk 6 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:124]: Ah, definitely should be, can check.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:124]: Meta-point: would be nice to enforce this, but I guess the easiest way to do so is just to have a non-infinite default step timeout.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/ReplicatedTearOffActor.cpp (hunk 7 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/ReplicatedTearOffActor.cpp:27]: Yeah, produces an error with any replication graph. Took some digging to find this.
Basically, when spawning a *dynamic* actor, the begin play is called immediately upon spawn, so we are still in the actor creation flow. It turns out actors are added to replication lists *just* after beginPlay is called on them. So tearing off an actor with replication graph will attempt to remove it from the replication list, but can't find it there, producing an error.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp (hunk 8 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp:89]: Yeah... but I don't like this entire situation. I guess ideally the test could be reworked so it uses the debug interface instead? That way we won't have to worry about things migrating at unfortunate times.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp:89]: I could add it to the ticket to fix this test, to consider reworking it to use the debug interface.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp (hunk 9 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp:85]: Ah, true, can move it.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp (hunk 10 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp:87]: Hm, I *think* the clients don't have the replication driver? So I have to populate this from a server to pass the information of "running with rep graph" down to the clients. I think.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp (hunk 4 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForInterestActor.cpp]: Cool - because the requires are reset each time they're called this is fine 👍

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/ReplicatedTearOffActor.cpp (hunk 7 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/ReplicatedTearOffActor.cpp:27]: Might be worth adding that to this comment - as it's useful context.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp (hunk 8 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp:89]: Sounds good

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp (hunk 10 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp]: Yeah you're right - can just check `ReplicationDriverClassName` though

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp]: Oh, that could be easier. Is that robust? I did this based on what we do in the NetDriver.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp (hunk 11 above)
Comment (jeffxu-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestInitialOnly/SpatialTestInitialOnlyForSpawnActor.cpp:108]: `RequireEqual` and `FinishStep`, cool. 
I got it. Thanks.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp (hunk 10 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp]: Yep should be - it's just reading a ini config,

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp]: Actually, what would happen if the replication graph class in the ini is not found?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (improbable-dmitrii) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1857]: Looks good

**Comment by mironec** [comment]
Another slow run: https://buildkite.com/improbable/unrealgdk-premerge/builds/23587

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp (hunk 10 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp]: But I changed it, it now uses the class name.

**Review by m-samiec — APPROVED** [review]
Nice work!
